File: length/length_test.go
package length_test

import (
	"testing"

	"github.com/retbrown/archeryutils/length"
)

func TestKnownUnits(t *testing.T) {
	want := map[string]struct{}{
		"cm": {}, "inch": {}, "metre": {}, "yard": {},
	}
	if len(length.KnownUnits) != len(want) {
		t.Fatalf("KnownUnits = %v, want %v", length.KnownUnits, want)
	}
	for k := range want {
		if _, ok := length.KnownUnits[k]; !ok {
			t.Errorf("KnownUnits missing %q", k)
		}
	}
}

func TestUnitSetsContainDefinitiveName(t *testing.T) {
	tests := []struct {
		name string
		set  length.UnitSet
	}{
		{"cm", length.CM},
		{"inch", length.Inch},
		{"metre", length.Metre},
		{"yard", length.Yard},
	}
	for _, tc := range tests {
		if _, ok := tc.set[tc.name]; !ok {
			t.Errorf("set %q does not contain definitive name %q", tc.name, tc.name)
		}
	}
}

func TestUnitSetsContainPlurals(t *testing.T) {
	tests := []struct {
		plural string
		set    length.UnitSet
	}{
		{"cms", length.CM},
		{"inches", length.Inch},
		{"metres", length.Metre},
		{"yards", length.Yard},
	}
	for _, tc := range tests {
		if _, ok := tc.set[tc.plural]; !ok {
			t.Errorf("set does not contain plural %q", tc.plural)
		}
	}
}

func TestToMetres(t *testing.T) {
	tests := []struct {
		value float64
		unit  string
		want  float64
	}{
		{10, "metre", 10},
		{10, "cm", 0.1},
		{10, "inch", 0.254},
		{10, "yard", 9.144},
	}
	for _, tc := range tests {
		got, err := length.ToMetres(tc.value, tc.unit)
		if err != nil {
			t.Errorf("ToMetres(%v,%q) error: %v", tc.value, tc.unit, err)
			continue
		}
		if !approxEqual(got, tc.want, 1e-9) {
			t.Errorf("ToMetres(%v,%q) = %v, want %v", tc.value, tc.unit, got, tc.want)
		}
	}
}

func TestFromMetres(t *testing.T) {
	tests := []struct {
		value float64
		unit  string
		want  float64
	}{
		{10, "metre", 10},
		{10, "cm", 1000},
		{10, "inch", 393.7008},
		{10, "yard", 10.93613},
	}
	for _, tc := range tests {
		got, err := length.FromMetres(tc.value, tc.unit)
		if err != nil {
			t.Errorf("FromMetres(%v,%q) error: %v", tc.value, tc.unit, err)
			continue
		}
		if !approxEqual(got, tc.want, 1e-4) {
			t.Errorf("FromMetres(%v,%q) = %v, want %v", tc.value, tc.unit, got, tc.want)
		}
	}
}

func TestDefinitiveUnit(t *testing.T) {
	tests := []struct {
		alias string
		want  string
	}{
		{"m", "metre"},
		{"centimetre", "cm"},
		{"Inch", "inch"},
		{"yd", "yard"},
	}
	for _, tc := range tests {
		got, err := length.DefinitiveUnit(tc.alias)
		if err != nil {
			t.Errorf("DefinitiveUnit(%q) error: %v", tc.alias, err)
			continue
		}
		if got != tc.want {
			t.Errorf("DefinitiveUnit(%q) = %q, want %q", tc.alias, got, tc.want)
		}
	}
}

func TestDefinitiveUnits(t *testing.T) {
	set := length.Union(length.Inch, length.CM)
	got, err := length.DefinitiveUnits(set)
	if err != nil {
		t.Fatalf("DefinitiveUnits error: %v", err)
	}
	want := length.UnitSet{"inch": {}, "cm": {}}
	if len(got) != len(want) {
		t.Fatalf("DefinitiveUnits = %v, want %v", got, want)
	}
	for k := range want {
		if _, ok := got[k]; !ok {
			t.Errorf("DefinitiveUnits result missing %q", k)
		}
	}
}

func TestParseOptionalUnitsFloat(t *testing.T) {
	supported := length.Union(length.Metre, length.Yard)
	tests := []struct {
		name        string
		value       any
		wantVal     float64
		wantUnit    string
	}{
		{"int-scalar", 10, 10, "metre"},
		{"float-scalar", 10.1, 10.1, "metre"},
		{"default-units", [2]any{10, "Metres"}, 10, "metre"},
		{"other-units", [2]any{10, "yds"}, 10, "yard"},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			v, u, err := length.ParseOptionalUnitsFloat(tc.value, supported, "metre")
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if !approxEqual(v, tc.wantVal, 1e-12) {
				t.Errorf("value = %v, want %v", v, tc.wantVal)
			}
			if u != tc.wantUnit {
				t.Errorf("unit = %q, want %q", u, tc.wantUnit)
			}
		})
	}
}

func TestParseOptionalUnitsUnsupported(t *testing.T) {
	supported := length.Union(length.Metre, length.Yard)
	_, _, err := length.ParseOptionalUnitsFloat([2]any{10.0, "bannana"}, supported, "metre")
	if err == nil {
		t.Fatal("expected error for unsupported unit, got nil")
	}
}

func TestParseOptionalUnitsDefaultNotSupported(t *testing.T) {
	supported := length.Union(length.Metre, length.Yard)
	_, _, err := length.ParseOptionalUnitsFloat(10.0, supported, "inch")
	if err == nil {
		t.Fatal("expected error for unsupported default unit, got nil")
	}
}

func approxEqual(a, b, tol float64) bool {
	d := a - b
	if d < 0 {
		d = -d
	}
	return d <= tol
}

File: ts/src/length.test.ts
import { describe, it, expect } from 'vitest';
import { toMetres, fromMetres, definitiveUnit, SUPPORTED_DIAM_UNITS, SUPPORTED_DIST_UNITS } from './length.js';

describe('toMetres', () => {
  it('metres pass through unchanged', () => {
    expect(toMetres(10, 'metre')).toBe(10);
  });
  it('centimetres to metres', () => {
    expect(toMetres(10, 'cm')).toBeCloseTo(0.1, 9);
  });
  it('inches to metres', () => {
    expect(toMetres(10, 'inch')).toBeCloseTo(0.254, 9);
  });
  it('yards to metres', () => {
    expect(toMetres(10, 'yard')).toBeCloseTo(9.144, 9);
  });
  it('alias: Metres', () => {
    expect(toMetres(5, 'Metres')).toBe(5);
  });
  it('alias: Yard', () => {
    expect(toMetres(100, 'Yard')).toBeCloseTo(91.44, 6);
  });
  it('throws for unsupported unit', () => {
    expect(() => toMetres(1, 'furlong')).toThrow();
  });
});

describe('fromMetres', () => {
  it('metres pass through unchanged', () => {
    expect(fromMetres(10, 'metre')).toBe(10);
  });
  it('metres to centimetres', () => {
    expect(fromMetres(0.1, 'cm')).toBeCloseTo(10, 9);
  });
  it('metres to yards', () => {
    expect(fromMetres(9.144, 'yard')).toBeCloseTo(10, 9);
  });
  it('throws for unsupported unit', () => {
    expect(() => fromMetres(1, 'furlong')).toThrow();
  });
});

describe('definitiveUnit', () => {
  it('yard aliases resolve to yard', () => {
    for (const a of ['yard', 'Yard', 'yards', 'Yards', 'yd', 'Yd', 'Y', 'y']) {
      expect(definitiveUnit(a)).toBe('yard');
    }
  });
  it('metre aliases resolve to metre', () => {
    for (const a of ['metre', 'Metre', 'metres', 'Metres', 'm', 'M']) {
      expect(definitiveUnit(a)).toBe('metre');
    }
  });
  it('cm aliases resolve to cm', () => {
    for (const a of ['cm', 'CM', 'cms', 'centimetre', 'Centimetre']) {
      expect(definitiveUnit(a)).toBe('cm');
    }
  });
  it('inch aliases resolve to inch', () => {
    for (const a of ['inch', 'Inch', 'inches', 'Inches']) {
      expect(definitiveUnit(a)).toBe('inch');
    }
  });
  it('throws for unrecognised alias', () => {
    expect(() => definitiveUnit('furlong')).toThrow();
  });
});

describe('SUPPORTED_DIAM_UNITS', () => {
  it('includes cm, inch, metre', () => {
    expect(SUPPORTED_DIAM_UNITS.has('cm')).toBe(true);
    expect(SUPPORTED_DIAM_UNITS.has('inch')).toBe(true);
    expect(SUPPORTED_DIAM_UNITS.has('metre')).toBe(true);
  });
  it('does not include yard', () => {
    expect(SUPPORTED_DIAM_UNITS.has('yard')).toBe(false);
  });
});

describe('SUPPORTED_DIST_UNITS', () => {
  it('includes yard and metre', () => {
    expect(SUPPORTED_DIST_UNITS.has('yard')).toBe(true);
    expect(SUPPORTED_DIST_UNITS.has('metre')).toBe(true);
  });
  it('does not include cm or inch', () => {
    expect(SUPPORTED_DIST_UNITS.has('cm')).toBe(false);
    expect(SUPPORTED_DIST_UNITS.has('inch')).toBe(false);
  });
});
